ibazel/ibazel.go

```diff
@@ -322,15 +329,75 @@ func (i *IBazel) iteration(command string, commandToRun runnableCommand, targets
 			i.state = RUN
 		}
 	case RUN:
-		log.Logf("%s %s", strings.Title(verb(command)), joinedTargets)
-		i.beforeCommand(targets, command)
-		outputBuffer, err := commandToRun(targets...)
-		i.interruptCount = 0
-		i.afterCommand(targets, command, err == nil, outputBuffer)
-		i.state = WAIT
+		runDone := make(chan bool)
+		runFn := func(i *IBazel, runDone chan bool) {
+			log.Logf("%s %s", strings.Title(verb(command)), joinedTargets)
+			i.beforeCommand(targets, command)
+			outputBuffer, err := commandToRun(targets...)
+			i.interruptCount = 0
+			i.afterCommand(targets, command, err == nil, outputBuffer)
+			i.state = WAIT
+			runDone <- true
+		}
+
+		go runFn(i, runDone)
+
+		// use abort compilation early behaviour
+		if i.abortLoopEarly {
+			select {
+			case e := <-i.sourceFileWatcher.Events():
+				if _, ok := i.filesWatched[i.sourceFileWatcher][e.Name]; ok && e.Op&modifyingEvents != 0 {
+					log.Logf("Changed: %q. Rebuilding...", e.Name)
+					i.logRunAbort(e.Name, "source")
+					i.cancel()
+					i.changeDetected(targets, "source", e.Name)
+					i.state = DEBOUNCE_RUN
+				}
+			case e := <-i.buildFileWatcher.Events():
+				if _, ok := i.filesWatched[i.buildFileWatcher][e.Name]; ok && e.Op&modifyingEvents != 0 {
+					i.logRunAbort(e.Name, "build")
+					i.cancel()
+					i.changeDetected(targets, "graph", e.Name)
+					i.state = DEBOUNCE_QUERY
+				}
+			case <-runDone:
+			}
+
+			return
+		}
+
+		// use default compilation behaviour
+		<-runDone
+
 	}
 }
 
+func (i *IBazel) logRunAbort(errorName string, abortType string) {
```

Comment:
Once you make another state I think this logic will be better folded into those two states.